### PR TITLE
Seq2Seq finetuning

### PR DESCRIPTION
**Pull request by alextrott16**
This PR covers seq-to-seq finetuning (such as IFT) for LLMs. These changes create a new dataloading codepath. All major changes are within `llm/src/data/finetuning/`.

At a high level:
- `finetuning/collator.py` a one-size-fits-all (for single-turn) seq2seq collator that turns properly formatted, tokenized examples into batches suitable for training with any of our LLM models.
- `finetuning/tasks.py` provides a scaffold for registering functions to format+tokenize examples from different datasources.
- `finetuning/dataloader.py` exposes `build_finetuning_dataloader` and interprets the dataloader config.
- `finetuning/convert_finetuning_dataset.py` simplifies the conversion of HF datasets into an MDS format that the dataloader can build from.
- `finetuning/README.md` explains how to use these tools.

Other changes:
- Various `__init__`s to simplify importing
- `main.py` can call `build_finetuning_dataloader`
- Example finetuning YAMLs in `yamls/mosaic_gpt/finetuning/`
- Tests
- Improvements to `packing.py`

### CODE REVIEW

File: examples/llm/src/data/finetuning/README.md [hunk 1]
```diff
@@ -2,20 +2,93 @@
...
+train_loader:
+    name: finetuning
+    dataset:
+        name: my-finetuning-task
+        ...
+```
+
+**IMPORTANT:** The subfield `dataset.name` has a special meaning. It tells the dataloader what function to use when tokenizing each example into the input and output texts.
+- "input" refers to the text that you feed into the model, e.g. *Tell me a few facts about dogs.*
+- "output" refers to the text that the model is trained to produce in response to the input, e.g. *Dogs are great pets. They love to play fetch...*
```
Comment (samhavens) [inline examples/llm/src/data/finetuning/README.md:24]:
```suggestion
- "output" refers to the text that the model is trained to produce in response to the input, e.g. *Dogs are great pets. They love to play fetch. They are better than cats...*
```

evil grin

**Comment by alextrott16** [comment]
Note that this will need to be partially re-written after https://github.com/mosaicml/examples/pull/314 goes through.

**Review by dakinggg — APPROVED** [review]
I confess this was a lot of code and I may have missed some stuff...but it basically looks good to me. Seems easy to use, lots of nice functionality, cleanly incorporated.

File: examples/llm/src/data/finetuning/dataloader.py [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+
+    dataloader = build_finetuning_dataloader(cfg, device_batch_size)
+
+    tokenizer = transformers.AutoTokenizer.from_pretrained(
+        cfg.dataset.tokenizer_name)
+
+    packing = cfg.dataset.get('packing_ratio') is not None
+
+    for i, batch in enumerate(dataloader):
+        if i >= 100:
```
[the remark this answers is not in the document]
Comment (alextrott16) [reply examples/llm/src/data/finetuning/dataloader.py:214]: Ah, this was to let the dataloader run through enough batches to get good packing estimates. This is no longer needed because the packing profiling tooling has been added to the packing file itself.
I'll clean this up.

File: examples/llm/yamls/mosaic_gpt/finetune/1b_local_data_sft.yaml [hunk 3]
```diff
@@ -0,0 +1,128 @@
+tokenizer_name: gpt2 # This must match the tokenizer you pretrained with
```
Comment (vchiley) [inline examples/llm/yamls/mosaic_gpt/finetune/1b_local_data_sft.yaml:1]:
```suggestion
```

File: examples/llm/yamls/mosaic_gpt/finetune/1b_local_data_sft.yaml [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+# Run Name
+run_name: # If left blank, will be read from env var $COMPOSER_RUN_NAME
+
+# Model
+# These must match pretraining
+model:
+  name: mosaic_gpt
+  init_device: meta
+  tokenizer_name: ${tokenizer_name}
```
Comment (vchiley) [inline examples/llm/yamls/mosaic_gpt/finetune/1b_local_data_sft.yaml:13]:
```suggestion
```

File: examples/llm/yamls/mosaic_gpt/finetune/1b_local_data_sft.yaml [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+  n_heads: 16 # Modified 24->16 so that d_head == 128 to statisfy FlashAttention
+  n_layers: 24
+  mlp_ratio: 4
+  max_seq_len: ${max_seq_len}
+  vocab_size: 50368
+  attn_impl: triton
+
+# Tokenizer
+tokenizer:
+  name: ${tokenizer_name}
```
Comment (vchiley) [inline examples/llm/yamls/mosaic_gpt/finetune/1b_local_data_sft.yaml:24]:
```suggestion
  name: gpt2
```

File: examples/llm/yamls/mosaic_gpt/finetune/1b_local_data_sft.yaml [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+      train: /path/to/train.jsonl
+      test: /path/to/test.jsonl
+
+# Dataloaders
+train_loader: &train_loader
+  name: finetuning
+  dataset:
+    <<: *hf_dataset
+    split: train
+    tokenizer_name: ${tokenizer_name}
```
Comment (vchiley) [inline examples/llm/yamls/mosaic_gpt/finetune/1b_local_data_sft.yaml:42]:
```suggestion
```

File: examples/llm/yamls/mosaic_gpt/finetune/1b_local_data_sft.yaml [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+    shuffle: true
+  drop_last: true
+  num_workers: 8
+  pin_memory: false
+  prefetch_factor: 2
+  persistent_workers: true
+  timeout: 0
+
+eval_loader:
+  <<: *train_loader
```
Comment (vchiley) [inline examples/llm/yamls/mosaic_gpt/finetune/1b_local_data_sft.yaml:56]: what is this sorcery
❤️

File: examples/llm/yamls/mosaic_gpt/finetune/7b_dolly_sft.yaml [hunk 8]
```diff
@@ -0,0 +1,117 @@
+# data_local: ./my-copy-c4
+# data_remote: # If blank, files must be present in data_local
+tokenizer_name: gpt2
```
Comment (vchiley) [inline examples/llm/yamls/mosaic_gpt/finetune/7b_dolly_sft.yaml:3]:
```suggestion
```

File: examples/llm/yamls/mosaic_gpt/finetune/7b_dolly_sft.yaml [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+global_seed: 17
+
+# Run Name
+run_name: # If left blank, will be read from env var $COMPOSER_RUN_NAME
+
+# Model
+model:
+  name: mosaic_gpt
+  init_device: meta
+  tokenizer_name: ${tokenizer_name}
```
Comment (vchiley) [inline examples/llm/yamls/mosaic_gpt/finetune/7b_dolly_sft.yaml:14]:
```suggestion
```

File: examples/llm/yamls/mosaic_gpt/finetune/7b_dolly_sft.yaml [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+  n_heads: 32
+  n_layers: 32
+  mlp_ratio: 4
+  max_seq_len: ${max_seq_len}
+  vocab_size: 50368
+  attn_impl: triton
+
+# Tokenizer
+tokenizer:
+  name: ${tokenizer_name}
```
Comment (vchiley) [inline examples/llm/yamls/mosaic_gpt/finetune/7b_dolly_sft.yaml:25]:
```suggestion
  name: gpt2
```

File: examples/llm/yamls/mosaic_gpt/finetune/7b_dolly_sft.yaml [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+  kwargs:
+    model_max_length: ${max_seq_len}
+
+# Dataloaders
+train_loader:
+  name: finetuning
+  dataset:
+    name: HuggingFaceH4/databricks_dolly_15k
+    split: train
+    tokenizer_name: ${tokenizer_name}
```
Comment (vchiley) [inline examples/llm/yamls/mosaic_gpt/finetune/7b_dolly_sft.yaml:35]:
```suggestion
```

**Review by vchiley — COMMENTED** [review]
for tok updt

(you don't need to actually apply, the changes here, but I wanted to note them all)

File: examples/llm/src/data/finetuning/dataloader.py [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+                print(k, v.shape)
+            else:
+                print(k, v)
+        for j in range(device_batch_size):
+            print(f'--- Sample {j} ---')
+            if cfg.dataset.decoder_only_format:
+                if packing:
+                    for subseq in range(int(batch['sequence_id'][j].max()) + 1):
+                        is_subseq = batch['sequence_id'][j] == subseq
+                        print(
```
[the remark this answers is not in the document]
Comment (alextrott16) [reply examples/llm/src/data/finetuning/dataloader.py:232]: Oh, it works.

File: examples/llm/yamls/mosaic_gpt/finetune/1b_local_data_sft.yaml (hunk 7 above)
Comment (alextrott16) [reply examples/llm/yamls/mosaic_gpt/finetune/1b_local_data_sft.yaml:56]: @samhavens has discovered some YAML dark magics. Life will never be the same.

Comment (samhavens) [reply examples/llm/yamls/mosaic_gpt/finetune/1b_local_data_sft.yaml:56]: Yes! We already are using YAML anchors and aliases for strings, but it turns out you can use them for dicts and lists as well!
https://joshdevlin.com/blog/yaml-repeating-sections/

Comment (samhavens) [reply examples/llm/yamls/mosaic_gpt/finetune/1b_local_data_sft.yaml:56]: actually here is a more authoritative source https://yaml.org/type/merge.html
